File: typescript/src/acquisition.ts
import { DuplicateSearchError, ProtocolContractError } from "./errors.js";

export interface SearchController {
  searchRaw(arguments_: Record<string, unknown>): Promise<unknown>;
}

/** One private controller search replaces, rather than duplicates, MCP search. */
export class ControllerTurn {
  readonly modelMcpTools = ["remember"] as const;
  readonly #controller: SearchController;
  #searched = false;

  constructor(controller: SearchController) {
    this.#controller = controller;
  }

  async searchRawOnce(arguments_: Record<string, unknown>): Promise<unknown> {
    if (this.#searched) {
      throw new DuplicateSearchError(
        "controller acquisition already replaced MCP search for this turn",
      );
    }
    this.#searched = true;
    return this.#controller.searchRaw(arguments_);
  }
}

/**
 * Select original, index-aligned refused items without retrying or returning
 * accepted writes. The caller repairs and explicitly resubmits these items.
 */
export function refusedBatchItems(
  arguments_: Readonly<Record<string, unknown>>,
  response: Readonly<Record<string, unknown>>,
): Readonly<Record<string, unknown>>[] {
  const items = arguments_.items;
  const results = response.results;
  if (!Array.isArray(items) || !Array.isArray(results) || items.length !== results.length) {
    throw new ProtocolContractError(
      "remember batch response must align one result with each item",
    );
  }
  const refused: Readonly<Record<string, unknown>>[] = [];
  for (let index = 0; index < items.length; index += 1) {
    const item = items[index];
    const result = results[index];
    if (!isRecord(item) || !isRecord(result)) {
      throw new ProtocolContractError("remember batch items and results must be JSON objects");
    }
    if (typeof result.status !== "string") {
      throw new ProtocolContractError("remember batch result status must be a string");
    }
    if (result.status === "refused") refused.push(item);
  }
  return refused;
}

function isRecord(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: python/src/kaleidoscope_memory/acquisition.py
"""Controller-owned turn guard: native acquisition replaces the MCP search."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any, Mapping, Protocol

from .errors import DuplicateSearchError, ProtocolError


class SearchController(Protocol):
    async def search_raw(self, arguments: Mapping[str, Any]) -> Any: ...


class ControllerTurn:
    """Allow exactly one private search and remove search from model tools."""

    model_mcp_tools = ("remember",)

    def __init__(self, controller: SearchController) -> None:
        self._controller = controller
        self._searched = False

    async def search_raw_once(self, arguments: Mapping[str, Any]) -> Any:
        if self._searched:
            raise DuplicateSearchError("controller acquisition already replaced MCP search for this turn")
        self._searched = True
        return await self._controller.search_raw(arguments)


def refused_batch_items(
    arguments: Mapping[str, Any], response: Mapping[str, Any]
) -> list[Mapping[str, Any]]:
    """Return the original items refused by an index-aligned batch response.

    This is intentionally a pure selector, not an automatic retry. The caller
    must repair the returned items and explicitly choose whether to resubmit
    them; accepted items are never returned or duplicated.
    """

    items = arguments.get("items")
    results = response.get("results")
    if (
        not isinstance(items, Sequence)
        or isinstance(items, (str, bytes, bytearray))
        or not isinstance(results, Sequence)
        or isinstance(results, (str, bytes, bytearray))
        or len(items) != len(results)
    ):
        raise ProtocolError("remember batch response must align one result with each item")

    refused: list[Mapping[str, Any]] = []
    for item, result in zip(items, results, strict=True):
        if not isinstance(item, Mapping) or not isinstance(result, Mapping):
            raise ProtocolError("remember batch items and results must be JSON objects")
        status = result.get("status")
        if not isinstance(status, str):
            raise ProtocolError("remember batch result status must be a string")
        if status == "refused":
            refused.append(item)
    return refused
